### ISSUE TITLE
3.14 annotation evaluation is deferred

### ISSUE THREAD

**Issue by nickdrozd** [pylint-dev/pylint#10149]
### Bug description

```python
In 3.14, evaluation of annotations is deferred. Pylint raises undefined-variable for the field annotation below, but in 3.14 this is a false positive.

class X:
    x: X  # <-- undefined-variable
```

### Configuration

```ini

```

### Command used

```shell
pylint annotation.py
```

### Pylint output

```python
annotation.py:2:7: E0602: Undefined variable 'X' (undefined-variable)
```

### Expected behavior

no warning

### Pylint version

```shell
pylint 3.3.2
astroid 3.3.6
Python 3.14.0a3+ (heads/main:2a66dd33dfc, Dec 21 2024, 16:04:14) [GCC 9.4.0]
```

### Additional dependencies

```python

```

**Comment by nickdrozd** [comment]
Also `used-before-assignment` false positives

**Comment by nickdrozd** [comment]
Python 3.14 is now in beta. The new annotation behavior is handled correctly by Mypy and Ruff.

**Comment by Pierre-Sassoulas** [comment]
Do you want to contribute Nick ? Now that the beta is out and available in github actions, there's https://github.com/pylint-dev/astroid/pull/2731 on the astroid side but you can already work on the pylint side even if astroid is not officially 3.14 ready yet :) 

We do not work on new features during alphas because we already did in the past and then had to revert some of the things we did before the beta was released. ruff is supported by astral and mypy is supported by dropbox so they can probably afford to make different choices. pylint maintenance is volunteer based (except for CVE, supported by Tidelift), so if no one is intrinsically motivated to do something nothing happen.

**Comment by nickdrozd** [comment]
No, it seems like it will be a lot of work. I am happy to help with PR review though.

Probably it is going to require a complete overhaul of the variables checker. I think it would be a good idea to fix this issue and https://github.com/pylint-dev/pylint/issues/9815 and other variable false positives (walrus, etc) all at once.

**Comment by jacobtylerwalls** [comment]
I think the patch is just:

```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index 32ec31809..e8d12eebb 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1363,7 +1363,11 @@ class VariablesChecker(BaseChecker):
         checks globals doesn't overrides builtins.
         """
         self._to_consume = [NamesConsumer(node, "module")]
-        self._postponed_evaluation_enabled = is_postponed_evaluation_enabled(node)
+        py_version = self.linter.config.py_version
+        py314_plus = py_version >= (3, 14)
+        self._postponed_evaluation_enabled = (
+            py314_plus or is_postponed_evaluation_enabled(node)
+        )
 
         for name, stmts in node.locals.items():
             if utils.is_builtin(name):
```
